### PR TITLE
Ovs upstream for review2

### PR DESCRIPTION
**Pull request by OphirMunk**

### CODE REVIEW

File: lib/netdev-rte-offloads.c [hunk 1]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
     return 0;
 }
 
+struct ufid_to_odp {
+    struct cmap_node node;
+    ovs_u128 ufid;
+    odp_port_t port_no;
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: rename port_no

File: lib/netdev-rte-offloads.c [hunk 2]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+
+static struct cmap ufid_to_portid_map = CMAP_INITIALIZER;
+
+/*
+ * Search for ufid mapping
+ *
+ * Return ref to object and not a copy.
+ */
+static struct ufid_to_odp *
+ufid_to_portid_get(const ovs_u128 * ufid)
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: remove space "* ufid"

File: lib/netdev-rte-offloads.c [hunk 3]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+{
+    size_t hash = hash_bytes(ufid, sizeof(*ufid), 0);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: sizeof should be without ()

File: lib/netdev-rte-offloads.c [hunk 4]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+    struct ufid_to_odp * data;
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: remove space

File: lib/netdev-rte-offloads.c [hunk 5]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+        }
+    }
+
+    return NULL;
+}
+
+static odp_port_t
+ufid_to_portid_search(const ovs_u128 * ufid)
+{
+   struct ufid_to_odp * data = ufid_to_portid_get(ufid);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: remove space

File: lib/netdev-rte-offloads.c [hunk 6]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+
+   return (data) ? data->port_no : INVALID_ODP_PORT;
+}
+
+/*
+ * Save the fuid->port_no mapping.
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: typo fuid

File: lib/netdev-rte-offloads.c [hunk 7]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+ *
+ * Return the port if saved successfully.
+ */
+static odp_port_t
+ufid_to_portid_add(const ovs_u128 *ufid, odp_port_t port_no) OVS_UNUSED;
+static odp_port_t
+ufid_to_portid_add(const ovs_u128 * ufid, odp_port_t port_no)
+{
+    size_t hash = hash_bytes(ufid, sizeof(*ufid), 0);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: sizeof

File: lib/netdev-rte-offloads.c [hunk 8]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+    struct ufid_to_odp *data;
+
+    if (ufid_to_portid_search(ufid) != INVALID_ODP_PORT) {
+        return port_no;
+    }
+
+    data = xzalloc(sizeof(*data));
+
+    if (data == NULL) {
+        VLOG_WARN("failed to add fuid to odp, OOM");
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: typo fuid

File: lib/netdev-rte-offloads.c [hunk 9]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+
+/*
+ * Save the fuid->port_no mapping.
+ *
+ * Return the port if saved successfully.
+ */
+static odp_port_t
+ufid_to_portid_add(const ovs_u128 *ufid, odp_port_t port_no) OVS_UNUSED;
+static odp_port_t
+ufid_to_portid_add(const ovs_u128 * ufid, odp_port_t port_no)
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: remove space after *

File: lib/netdev-rte-offloads.c [hunk 10]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+                CONST_CAST(struct cmap_node *, &data->node), hash);
+
+    return port_no;
+}
+
+/*
+ * Remove the mapping if exists.
+ */
+static void
+ufid_to_portid_remove(const ovs_u128 * ufid) OVS_UNUSED;
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: remove space after *

File: lib/netdev-rte-offloads.c [hunk 11]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+static void
+ufid_to_portid_remove(const ovs_u128 * ufid)
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: remove space after *

File: lib/netdev-rte-offloads.c [hunk 12]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+{
+    size_t hash = hash_bytes(ufid, sizeof(*ufid), 0);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: sizeof

File: lib/netdev-rte-offloads.c [hunk 13]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+        if (ovs_u128_equals(*ufid, data->ufid)) {
+            return data;
+        }
+    }
+
+    return NULL;
+}
+
+static odp_port_t
+ufid_to_portid_search(const ovs_u128 * ufid)
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: remove space after *

File: lib/netdev-rte-offloads.c [hunk 14]
```diff
@@ -105,6 +106,149 @@ netdev_rte_port_set(struct netdev *netdev, odp_port_t dp_port,
...
+    int max_flows;
+    int curr_idx;
+    struct rte_flow_params {
+        struct rte_flow *flow;
+        struct netdev *netdev;
+    } rte_flow_data[0];
+};
+
+/*
+ * fuid hw offload struct contains array of pointers to rte flows.
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: typo fuid

File: lib/netdev-rte-offloads.c [hunk 15]
```diff
@@ -105,6 +106,149 @@ netdev_rte_port_set(struct netdev *netdev, odp_port_t dp_port,
...
+ * There may be a one OVS flow to many rte flows. For example in case
+ * of vxlan OVS flow we add an rte flow per each phsical port.
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: typo phsical

File: lib/netdev-rte-offloads.c [hunk 16]
```diff
@@ -105,6 +106,149 @@ netdev_rte_port_set(struct netdev *netdev, odp_port_t dp_port,
...
+ *
+ * max_flows - number of expected max rte flows for this ufid.
+ * ufid - the ufid.
+ *
+ * Return allocated struct ufid_hw_offload or NULL if allocation failed.
+ */
+static struct ufid_hw_offload *
+netdev_rte_port_ufid_hw_offload_alloc(int max_flows,
+                                      const ovs_u128 * ufid) OVS_UNUSED;
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: remove space after *

File: lib/netdev-rte-offloads.c [hunk 17]
```diff
@@ -105,6 +106,149 @@ netdev_rte_port_set(struct netdev *netdev, odp_port_t dp_port,
...
+static struct ufid_hw_offload *
+netdev_rte_port_ufid_hw_offload_alloc(int max_flows, const ovs_u128 * ufid)
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: remove space after *

File: lib/netdev-rte-offloads.c [hunk 18]
```diff
@@ -105,6 +106,149 @@ netdev_rte_port_set(struct netdev *netdev, odp_port_t dp_port,
...
+{
+    struct ufid_hw_offload *ufidol =
+        xzalloc(sizeof(struct ufid_hw_offload) +
+                       sizeof(struct rte_flow_params) * max_flows);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: both sizeof should be without ()

File: lib/netdev-rte-offloads.c [hunk 19]
```diff
@@ -105,6 +106,149 @@ netdev_rte_port_set(struct netdev *netdev, odp_port_t dp_port,
...
+
+/*
+ * Given ufid find its hw_offload struct.
+ *
+ * Return struct ufid_hw_offload or NULL if not found.
+ */
+static struct ufid_hw_offload *
+ufid_hw_offload_find(const ovs_u128 *ufid, struct cmap *map)
+{
+    size_t hash = hash_bytes(ufid, sizeof(*ufid), 0);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: sizeof

File: lib/netdev-rte-offloads.c [hunk 20]
```diff
@@ -105,6 +106,149 @@ netdev_rte_port_set(struct netdev *netdev, odp_port_t dp_port,
...
+
+    return NULL;
+}
+
+static struct ufid_hw_offload *
+ufid_hw_offload_remove(const ovs_u128 *ufid, struct cmap *map) OVS_UNUSED;
+static struct ufid_hw_offload *
+ufid_hw_offload_remove(const ovs_u128 *ufid, struct cmap *map)
+{
+    size_t hash = hash_bytes(ufid, sizeof(*ufid), 0);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: sizeof

File: lib/netdev-rte-offloads.c [hunk 21]
```diff
@@ -105,6 +106,149 @@ netdev_rte_port_set(struct netdev *netdev, odp_port_t dp_port,
...
+ *
+ */
+static int netdev_rte_port_ufid_hw_offload_free(
+    struct ufid_hw_offload *hw_offload) OVS_UNUSED;
+static int
+netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
+{
+    struct rte_flow_error error;
+
+    VLOG_DBG("clean all rte flows for fuid "UUID_FMT".\n",
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: typo fuid

File: lib/dpif-netdev.c [hunk 22]
```diff
@@ -1867,7 +1867,7 @@ dpif_netdev_port_add(struct dpif *dpif, struct netdev *netdev,
         *port_nop = port_no;
         error = do_add_port(dp, dpif_port, netdev_get_type(netdev), port_no);
 #ifdef DPDK_NETDEV
-        netdev_rte_offloads_port_add(netdev);
+        netdev_rte_offloads_port_add(netdev, port_no);
```
Comment (asafpenso) [inline lib/dpif-netdev.c]: rename port_no

File: lib/netdev-rte-offloads.c (hunk 18 above)
Comment (OphirMunk) [reply lib/netdev-rte-offloads.c]: When using sizeof structs you must use paranthesis

**Comment by OphirMunk** [comment]
PR updated with review comments

File: lib/netdev-rte-offloads.c (hunk 2 above)
Comment (asafpenso) [reply lib/netdev-rte-offloads.c]: still need to handle

File: lib/netdev-rte-offloads.c (hunk 8 above)
Comment (asafpenso) [reply lib/netdev-rte-offloads.c]: still need to handle

File: lib/netdev-rte-offloads.c [hunk 23]
```diff
@@ -249,6 +251,94 @@ netdev_rte_port_ufid_hw_offload_free(struct ufid_hw_offload *hw_offload)
...
+static odp_port_t
+ufid_to_portid_search(const ovs_u128 *ufid)
+{
+   struct ufid_to_odp *data = ufid_to_portid_get(ufid);
+
+   return (data) ? data->dp_port : INVALID_ODP_PORT;
+}
+
+/*
+ * Save the fuid->dp_port mapping.
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: typo

File: lib/netdev-rte-offloads.c (hunk 4 above)
Comment (asafpenso) [reply lib/netdev-rte-offloads.c]: still need to handle

**Comment by yuskul** [comment]
bot:retest

File: lib/netdev-rte-offloads.c [hunk 24]
```diff
@@ -28,8 +28,83 @@
...
+    uint32_t exception_mark; /* Exception SW handling for this port type. */
+};
+
+static struct cmap port_map = CMAP_INITIALIZER;
+
+/*
+ * Search for offloaded port data by dp_port no.
+ */
+static struct netdev_rte_port *
+netdev_rte_port_search(odp_port_t dp_port, struct cmap *map)
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: Probably need to rename all "rte" functions to "dpdk"

File: lib/netdev-rte-offloads.c [hunk 25]
```diff
@@ -1021,38 +1000,85 @@ netdev_rte_offloads_flow_put(struct netdev *netdev, struct match *match,
...
 
-    return netdev_rte_offloads_add_flow(netdev, match, actions,
-                                        actions_len, ufid, info);
+    rte_flow = netdev_rte_offloads_add_flow(netdev, match, actions,
+                                            actions_len, ufid, info);
+    if (!rte_flow) {
+        return ENODEV;
+    }
+
+    ufid_hw_offload_add_rte_flow(ufid_hw_offload, rte_flow, netdev);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: Also in my code there's a debug print which is useful:
VLOG_DBG("%s: installed flow %p by ufid "UUID_FMT"\n",
        netdev_get_name(netdev), rte_flow, UUID_ARGS((struct uuid *)ufid));
What do you think?

File: lib/netdev-rte-offloads.c [hunk 26]
```diff
@@ -790,6 +790,37 @@ add_flow_patterns(struct flow_patterns *patterns,
...
+                                struct rte_flow_action_jump *jump,
+                                struct flow_actions *actions)
+{
+    odp_port_t odp_port;
+    struct netdev_rte_port *rte_port;
+
+    odp_port = nl_attr_get_odp_port(nlattr);
+    rte_port = netdev_rte_port_search(odp_port, &port_map);
+    if (!rte_port) {
+        VLOG_ERR("No rte port was found for odp_port %u",
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: I changed to DBG, since it's not an error always

File: lib/netdev-rte-offloads.c [hunk 27]
```diff
@@ -790,6 +790,37 @@ add_flow_patterns(struct flow_patterns *patterns,
...
+                odp_to_u32(odp_port));
+        return NULL;
+    }
+
+    jump->group = rte_port->table_id; // OMREVIEW - check if table_id is assigned. Maybe can be replaced with a macro definition
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: yes, table id is assigned and is used here. I'm removing the comment.

File: lib/netdev-rte-offloads.c [hunk 28]
```diff
@@ -993,7 +1078,7 @@ netdev_rte_offloads_flow_put(struct netdev *netdev, struct match *match,
 
 int
 netdev_rte_offloads_flow_del(struct netdev *netdev OVS_UNUSED,
-                             const ovs_u128 *ufid,
+                             const ovs_u128 *ufid,  // OMTODO - to check: how comes netdev is UNUSED???
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: Inside netdev_rte_port_ufid_hw_offload_free we call for netdev_dpdk_rte_flow_destroy for all netdevs. This is why the parameter netdev is now used.

File: lib/netdev-rte-offloads.c [hunk 29]
```diff
@@ -1021,11 +1106,96 @@ netdev_rte_offloads_flow_del(struct netdev *netdev OVS_UNUSED,
     return 0;
 }
 
+static int
+netdev_rte_vport_flow_del(struct netdev *netdev OVS_UNUSED,
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: Now that there is a single cmpa for dpdk and vport, the "del" functions are exactly the same. so i created another static function netdev_offloads_flow_del with the implementation which both del functions call it

File: lib/netdev-rte-offloads.c [hunk 30]
```diff
@@ -1021,11 +1106,96 @@ netdev_rte_offloads_flow_del(struct netdev *netdev OVS_UNUSED,
...
+                                   struct netdev_rte_port *rte_port OVS_UNUSED,
+                                   struct netdev *netdev OVS_UNUSED,
+                                   struct match *match OVS_UNUSED,
+                                   struct nlattr *nl_actions OVS_UNUSED,
+                                   size_t actions_len OVS_UNUSED,
+                                   const ovs_u128 *ufid OVS_UNUSED,
+                                   struct offload_info *info OVS_UNUSED,
+                                   struct dpif_flow_stats *stats OVS_UNUSED)
+{
+    return -1;
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: changed to value from errno

File: lib/netdev-rte-offloads.c [hunk 31]
```diff
@@ -1021,11 +1106,96 @@ netdev_rte_offloads_flow_del(struct netdev *netdev OVS_UNUSED,
...
+                 const ovs_u128 *ufid,
+                 struct offload_info *info,
+                 struct dpif_flow_stats *stats)
+{
+    if (netdev_rte_offloads_validate_flow(match, true)) { // OMTODO - consider renaming this API w/o "offloads"
+        VLOG_DBG("flow pattern not supported");
+        return -1;
+    }
+
+    int ret = 0;
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: changed to errno value

File: lib/netdev-rte-offloads.c [hunk 32]
```diff
@@ -1021,11 +1106,96 @@ netdev_rte_offloads_flow_del(struct netdev *netdev OVS_UNUSED,
...
+static int
+netdev_rte_vport_flow_put(struct netdev *netdev,
+                 struct match *match,
+                 struct nlattr *actions,
+                 size_t actions_len,
+                 const ovs_u128 *ufid,
+                 struct offload_info *info,
+                 struct dpif_flow_stats *stats)
+{
+    if (netdev_rte_offloads_validate_flow(match, true)) { // OMTODO - consider renaming this API w/o "offloads"
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: this function is already in upstream, so we can wait for Ilya change from "rte" to "dpdk" and have this comment there

File: lib/netdev-rte-offloads.c [hunk 33]
```diff
@@ -1021,11 +1106,96 @@ netdev_rte_offloads_flow_del(struct netdev *netdev OVS_UNUSED,
...
+                       ret = -1;
+                }
+                break;
+            case RTE_PORT_TYPE_DPDK:
+                VLOG_WARN("offload of vport could on dpdk port");
+                ret = -1;
+                break;
+            default:
+                VLOG_DBG("unsupported tunnel type"); // OMTODO - seems like an internal design error. Why not assert?
+                ret = -1;
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: removed ret and break and added ovs_assert(true);

File: lib/netdev-rte-offloads.c [hunk 34]
```diff
@@ -1021,11 +1106,96 @@ netdev_rte_offloads_flow_del(struct netdev *netdev OVS_UNUSED,
...
+    odp_port_t in_port = match->flow.in_port.odp_port;
+    struct netdev_rte_port *rte_port = netdev_rte_port_search(in_port,
+                                                              &port_map);
+// OMREVIEW: vport_map -> port_map No need for the checking below
+    if (rte_port != NULL) {
+        switch (rte_port->rte_port_type) {
+            case RTE_PORT_TYPE_VXLAN:
+                VLOG_DBG("vxlan offload ufid "UUID_FMT" \n",
+                         UUID_ARGS((struct uuid *)ufid));
+                if (netdev_vport_vxlan_add_rte_flow_offload(rte_port,
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: instead of "if" just did ret = netdev_vport_xxxx

File: lib/netdev-rte-offloads.c [hunk 35]
```diff
@@ -1021,11 +1106,96 @@ netdev_rte_offloads_flow_del(struct netdev *netdev OVS_UNUSED,
...
+    if (netdev_rte_offloads_validate_flow(match, true)) { // OMTODO - consider renaming this API w/o "offloads"
+        VLOG_DBG("flow pattern not supported");
+        return -1;
+    }
+
+    int ret = 0;
+    odp_port_t in_port = match->flow.in_port.odp_port;
+    struct netdev_rte_port *rte_port = netdev_rte_port_search(in_port,
+                                                              &port_map);
+// OMREVIEW: vport_map -> port_map No need for the checking below
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: will remove the switch and check only for VXLAN. for other cases assert.

File: lib/netdev-rte-offloads.c [hunk 36]
```diff
@@ -938,6 +997,36 @@ netdev_rte_offloads_add_flow(struct netdev *netdev,
         /* Actions are supported, offload the flow */
         flow = netdev_rte_offload_flow(netdev, info, &patterns, &actions,
                                        &flow_attr);
+        if (flow) {
+           /* odp_port_t port_id = ufid_to_portid_search(ufid);
+            struct netdev_rte_port *rte_port =
+                netdev_rte_port_search_by_port_no(port_id);*/
+
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: I understand both suggestions, can you elaborate on their advantages?

File: lib/netdev-rte-offloads.c [hunk 37]
```diff
@@ -938,6 +997,36 @@ netdev_rte_offloads_add_flow(struct netdev *netdev,
...
+                    struct rte_flow_error error;
+                    result = netdev_dpdk_rte_flow_destroy(rte_port->netdev,
+                                                          flow, &error);
+                    if (result != 0) {
+                        VLOG_ERR_RL(&error_rl, "rte flow destroy error: %u : message : "
+                                   "%s\n", error.type, error.message);
+                    }
+                    flow = NULL;
+                } else {
+                    rte_port->default_rte_flow[vport->table_id] = def_flow; // OMREVIEW - where is def_flow freed?
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: I added a new function to handle that: netdev_rte_port_del_default_rules

**Comment by OphirMunk** [comment]
Review done

**Pull request closed without merging**